### PR TITLE
Rewrite search algorithm

### PR DESCRIPTION
**Pull request by shuding**
This PR rewrites the search algorithm to use https://github.com/EthanRutherford/fast-fuzzy instead, the result improved a lot in a real and large Nextra site according to my test.

Also changed the option to `search` instead of `flexsearch` because we don't need to expose too much info about the underlying implementation.

## Examples

Here are some examples on nextra.site. Old / new:

### "meta": _meta should be the first result

CleanShot 2023-01-23 at 19 05 22@2x

### "chat": two results link to the same place

CleanShot 2023-01-23 at 19 15 00@2x

### CODE REVIEW

File: packages/nextra-theme-docs/src/components/anchor.tsx [hunk 1]
```diff
@@ -41,7 +41,7 @@ export const Anchor = forwardRef<HTMLAnchorElement, AnchorProps>(function (
     )
   }
 
-  if (nextVersion > 12 || config.newNextLinkBehavior) {
+  if (nextVersion > 12 || config.frontMatter.newNextLinkBehavior) {
```
Comment (B2o5T) [inline packages/nextra-theme-docs/src/components/anchor.tsx:44]: this is not in frontMatter https://github.com/shuding/nextra/blob/1e26811bdd4f0e6ab571f93497ef116c06fee06c/packages/nextra-theme-docs/src/contexts/config.tsx#L108-L110, we can remove it in v3

File: packages/nextra-theme-docs/src/contexts/config.tsx [hunk 2]
```diff
@@ -13,11 +13,15 @@ import {
...
-type Config<FrontMatterType = FrontMatter> = DocsThemeConfig &
-  Pick<
-    PageOpts<FrontMatterType>,
-    'flexsearch' | 'newNextLinkBehavior' | 'title' | 'frontMatter'
+type Config<FrontMatterType = FrontMatter> = DocsThemeConfig & {
+  title: PageOpts<FrontMatterType>['title']
+  frontMatter: Partial<
+    Pick<
+      PageOpts<FrontMatterType>,
+      'search' | 'newNextLinkBehavior' | 'title' | 'frontMatter'
```
[the remark this answers is not in the document]
Comment (B2o5T) [reply packages/nextra-theme-docs/src/contexts/config.tsx:21]: also, it's strange because you pick `frontMatter` for `frontMatter` property 🤔

**Comment by shuding** [comment]
I shouldn't have changed the frontmatter typing there... totally unintended 😬. The problem was that we had both `search` in frontmatter and `flexsearch` in nextra config, and I changed `flexsearch` to `search` and then they conflict...

**Comment by leeoniya** [comment]
probably a bit late for a pitch here, but fast-fuzzy ain't all that fast, and probably a bit too fuzzy :)

https://leeoniya.github.io/uFuzzy/demos/compare.html?libs=uFuzzy,fast-fuzzy&search=super%20ma

https://github.com/leeoniya/uFuzzy#a-biased-appraisal-of-similar-work

**Comment by B2o5T** [comment]
@leeoniya late is better than never :))

**Comment by leeoniya** [comment]
> I want to find a good algorithm to give scores to each page.

it sounds like you want something closer to a document/fulltext search that can also take into account term frequency and which part of the document it occurs in. while uFuzzy is a great fuzzy search, it would take more legwork for it to rank documents with those metrics. existing document searches (FlexSearch, Lyra) use an index, which gets big in memory and expensive to start up if you want typo tolerance and partial matching.

it would be interesting to experiment with a uFuzzy wrapper than can do reasonably well without an index by just doing a pre-pass over all facets together (pageTitle, title, content) to quicky reduce the search space and then do each facet and needle word independently to count total matches and have that as part of final ranking/sorting.

**Comment by nihgwu** [comment]
@shuding are you planning to ship this? I guess it will solve my problem as well
image

**Comment by zburk** [comment]
Also curious if there are any updates on merging this PR? Fixing the "two results link to the same place" helps clean up the search results a ton.

**Comment by dimaMachina** [comment]
Closed by https://github.com/shuding/nextra/pull/2569

**Pull request closed without merging**